File: backend/app/schemas/block.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

BlockType = Literal[
    "heading",
    "paragraph",
    "question",
    "answer",
    "quote",
    "highlight",
    "bullet_list",
    "numbered_list",
    "checklist",
    "divider",
    "image",
    "callout",
    "table",
    "guest_info",
    "content_idea",
    "personal_note",
]


class BlockBase(BaseModel):
    type: BlockType
    content: dict[str, Any] = Field(default_factory=dict)
    position: float = 0
    is_important: bool = False
    is_potential_content: bool = False
    linked_question_id: uuid.UUID | None = None


class BlockCreate(BlockBase):
    pass


class BlockUpdate(BaseModel):
    type: BlockType | None = None
    content: dict[str, Any] | None = None
    position: float | None = None
    is_important: bool | None = None
    is_potential_content: bool | None = None
    linked_question_id: uuid.UUID | None = None


class BlockResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    page_id: uuid.UUID
    type: BlockType
    content: dict[str, Any]
    position: float
    is_important: bool
    is_potential_content: bool
    linked_question_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/services/notebook/types/block.ts
/**
 * Mirrors backend/app/schemas/block.py / app/models/block.py (BLOCK_TYPES)
 * exactly for the type enum and top-level fields.
 *
 * `content` is untyped JSONB on the backend (`dict[str, Any]`) with no
 * server-side shape enforcement per type - there is no existing persisted
 * convention to match, so the shapes below are this frontend's own minimal
 * convention (see services/notebook/lib/block-content.ts for safe readers).
 *
 * "image" and "table" are part of the backend's BLOCK_TYPES enum but are
 * intentionally NOT exposed as creatable types here - there is no upload
 * pipeline or structured table editor backing them yet, and shipping a
 * block type with no reliable way to fill it in would be a fake feature.
 * See the feature's final report for details.
 */
export type BlockType =
  | "heading"
  | "paragraph"
  | "question"
  | "answer"
  | "quote"
  | "highlight"
  | "bullet_list"
  | "numbered_list"
  | "checklist"
  | "divider"
  | "callout"
  | "guest_info"
  | "content_idea"
  | "personal_note";

export type TextBlockContent = { text: string };
export type ListBlockContent = { items: string[] };
export type ChecklistItem = { text: string; checked: boolean };
export type ChecklistBlockContent = { items: ChecklistItem[] };

export type Block = {
  id: string;
  page_id: string;
  type: BlockType;
  content: Record<string, unknown>;
  position: number;
  is_important: boolean;
  is_potential_content: boolean;
  linked_question_id: string | null;
  created_at: string;
  updated_at: string;
};

export type BlockCreateInput = {
  type: BlockType;
  content: Record<string, unknown>;
  position: number;
  is_important?: boolean;
  is_potential_content?: boolean;
  linked_question_id?: string | null;
};

export type BlockUpdateInput = {
  type?: BlockType;
  content?: Record<string, unknown>;
  position?: number;
  is_important?: boolean;
  is_potential_content?: boolean;
  linked_question_id?: string | null;
};
